### ISSUE TITLE
halyard deploys new secrets without cleaning up old secrets

### ISSUE THREAD

**Issue by sdpoueme** [spinnaker/spinnaker#5968]
### Issue Summary:
Halyard is not cleaning up secrets generated during spinnaker config deployment. Raising this issue again in reference to #4420

### Cloud Provider(s):

Kubernetes

### Environment:
OpenStack

### Feature Area:
Halyard

### Description:
After running Spinnaker in production during 2 years, we are seeing a high number of secrets generated by `hal deploy apply` even though we use the delete orphaned command during deployments. It seems halyard is not cleaning up old secrets. 
`kubectl get secrets -n spinnaker | wc -l`
  `4419`

### Steps to Reproduce:
1. Create or update a standard spinnaker configuration
2. Run `hal deploy apply`
3. Look at the secrets created in the spinnaker namespace
4. Repeat and observe the growth in the number of secrets.

**Comment by ezimanyi** [comment]
@sdpoueme : Thanks for bringing this up again!  This is in general a problem that seems to arise a lot in Kubernetes (kubernetes/kubernetes#22368) (kubernetes/enhancements#948).

I'm not sure what a great solution would be from the standpoint of Halyard.  In general, Halyard will deploy a secret and a deployment pointing to it; the old secret needs to stay as long as any replica sets managed by the deployment still point to the old secret.  Maybe at the end of the deployment, Halyard could look for all secrets it deployed, and then see whether they are being used by any replica set?

I'm interested in this problem primarily from the perspective of how it affects [kleat](https://github.com/spinnaker/kleat), which we're building as a more Kubernetes-native way of deploying Spinnaker.  Kleat has the same issue; as it deploys with Kustomize, it will just create new secrets each time and they will build up unless cleaned up.

One potential solution there would be to use `kubectl apply --prune` with appropriate label selectors but that runs into issues ordering the deletes; you can't delete the secret until the deployment you updated rolls out to no longer depend on it [as noted in this KEP](kubernetes/enhancements#810).  Maybe in the `kleat` case a reasonable workaround is to do a `kubectl apply`, followed by a `kubectl apply --prune --labels ...` once the rollout is done (and we know we can safely delete anything unreferenced).

I think this will not be as easy with Halyard, as it doesn't give you the opportunity of seeing the YAML objects you're deploying before they get applied to the cluster.  Probably the best solution for Halyard would just be some separate command that looks for unreferenced items and deletes them...but then this wouldn't even need to be in Halyard and could probably be done manually/via a small script every now and then.

This isn't something that I'll be able to work on for Halyard as I'm focusing on `kleat` (though interested in the problem from that perspective).  (With one of the motivations of `kleat` being that Halyard is doing a lot of non-Kubernetes native things that make working on its Kubernetes deployments difficult.). I would probably suggest manually deleting the old secrets for now, while awaiting a better solution in `kleat`.

I'll leave this open and unassigned in case someone does want to work on this for Halyard, or in case people have workarounds/thoughts about this.

**Comment by sdpoueme** [comment]
Hi @ezimanyi  Thanks for the detailed feedback. I will share it with my team as well. We store our halyard configuration in Enterprise GitHub and after every `hal deploy apply`, we commit the changes (files generated) back to the repo. This is how we are able to keep rolling updates of Spinnaker. Some of these secret probably originate from the files that were generated by Halyard and because of how we are rolling updates, they are maintained across deployments.  I went through the RFC for `kleat` and I really like the fact that it is taking in consideration the need to have the halyard configuration in a git repo. Bashing into halyard to perform actions was not ideal for us, so we built a layer with groovy that allows us to execute halyard in a container when needed and mount the config as a volume in a workspace (after checking it out from Git). I agree with you conclusio - at this point there is no urgent need to fix this issue on halyard but it should definitely be taken in consideration for `kleat`.
